File: lib/services/api_config.dart
import 'dart:async';
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:network_info_plus/network_info_plus.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../utils/constants.dart';

/// Résout l'URL de base de l'API en fonction du réseau détecté (WiFi / connexion).
/// En priorité : backend local sur le même réseau, sinon production (Render).
class ApiConfig {
  ApiConfig._();
  static const _cacheKey = 'campuslink_api_base_url';
  static const _cacheTsKey = 'campuslink_api_base_url_ts';
  static const _cacheTtlSeconds = 300; // 5 min

  static final _dio = Dio(BaseOptions(
    connectTimeout: const Duration(seconds: 2),
    receiveTimeout: const Duration(seconds: 2),
  ));

  /// Construit la liste des URLs candidates : locales (selon IP WiFi) puis production.
  static Future<List<String>> _candidates() async {
    const production = 'https://campuslink-9knz.onrender.com/api';
    final list = <String>[];

    try {
      final info = NetworkInfo();
      final wifiIp = await info.getWifiIP();
      if (wifiIp != null &&
          wifiIp.isNotEmpty &&
          wifiIp != '0.0.0.0' &&
          !wifiIp.startsWith('127.')) {
        // Même sous-réseau : .1 (box), .125, .100, .50, .35
        final parts = wifiIp.split('.');
        if (parts.length == 4) {
          final prefix = '${parts[0]}.${parts[1]}.${parts[2]}.';
          for (final host in ['1', '2', '10', '35', '50', '100', '125', '127', '200', '254']) {
            list.add('http://${prefix}$host:8000/api');
          }
        }
      }
    } catch (_) {}

    // Émulateur Android
    list.add('http://10.0.2.2:8000/api');
    list.add(production);
    return list;
  }

  static Future<bool> _tryBaseUrl(String baseUrl) async {
    try {
      final url = baseUrl.endsWith('/') ? '${baseUrl}auth/login/' : '$baseUrl/auth/login/';
      final r = await _dio.get<dynamic>(
        url,
        options: Options(
          validateStatus: (s) => s == 200 || s == 405,
          receiveDataWhenStatusError: true,
        ),
      );
      // 200 = endpoint existe, 405 = GET non autorisé (login attend POST) → c'est notre API
      return r.statusCode == 200 || r.statusCode == 405;
    } catch (_) {
      return false;
    }
  }

  /// Résout l'URL de base : utilise le cache si valide, sinon détecte et met en cache.
  static Future<String> resolveBaseUrl() async {
    final prefs = await SharedPreferences.getInstance();
    final cached = prefs.getString(_cacheKey);
    final ts = prefs.getInt(_cacheTsKey);
    if (cached != null &&
        ts != null &&
        (DateTime.now().millisecondsSinceEpoch - ts) < _cacheTtlSeconds * 1000) {
      if (kDebugMode) {
        debugPrint('[ApiConfig] Using cached base URL: $cached');
      }
      return cached;
    }

    final candidates = await _candidates();
    for (final url in candidates) {
      if (await _tryBaseUrl(url)) {
        await prefs.setString(_cacheKey, url);
        await prefs.setInt(_cacheTsKey, DateTime.now().millisecondsSinceEpoch);
        if (kDebugMode) {
          debugPrint('[ApiConfig] Resolved base URL: $url');
        }
        return url;
      }
    }

    final fallback = AppConstants.apiBaseUrl;
    await prefs.setString(_cacheKey, fallback);
    await prefs.setInt(_cacheTsKey, DateTime.now().millisecondsSinceEpoch);
    if (kDebugMode) {
      debugPrint('[ApiConfig] Fallback to production: $fallback');
    }
    return fallback;
  }

  /// Invalide le cache (utile après changement de réseau).
  static Future<void> invalidateCache() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_cacheKey);
    await prefs.remove(_cacheTsKey);
  }
}

File: CampusExpo/campuslink-app/services/apiConfig.ts
/**
 * Résout l'URL de base de l'API selon le réseau (WiFi / connexion).
 * Essaie en priorité les backends locaux (liste de candidats courants), puis production.
 * Sans dépendance expo-network pour éviter les soucis d'installation.
 */
import axios from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';

const CACHE_KEY = 'campuslink_api_base_url';
const CACHE_TS_KEY = 'campuslink_api_base_url_ts';
const CACHE_TTL_MS = 5 * 60 * 1000; // 5 min

const PRODUCTION_URL = 'https://campuslink-9knz.onrender.com/api';

function isWeb(): boolean {
  // eslint-disable-next-line no-restricted-globals
  return typeof window !== 'undefined' && typeof document !== 'undefined';
}

/** Liste des URLs candidates : locales (émulateur + 192.168.x.x) puis production. */
function getCandidates(): string[] {
  return [
    'http://10.0.2.2:8000/api',
    'http://192.168.1.1:8000/api',
    'http://192.168.1.2:8000/api',
    'http://192.168.1.10:8000/api',
    'http://192.168.1.35:8000/api',
    'http://192.168.1.50:8000/api',
    'http://192.168.1.100:8000/api',
    'http://192.168.1.125:8000/api',
    'http://192.168.1.127:8000/api',
    'http://192.168.0.1:8000/api',
    'http://192.168.12.35:8000/api',
    PRODUCTION_URL,
  ];
}

async function tryBaseUrl(baseUrl: string): Promise<boolean> {
  try {
    const url = baseUrl.endsWith('/') ? `${baseUrl}auth/login/` : `${baseUrl}/auth/login/`;
    const r = await axios.get(url, {
      timeout: 3000,
      validateStatus: (s) => s === 200 || s === 405,
    });
    // 200 = endpoint existe, 405 = GET non autorisé (login attend POST) → c'est notre API
    return r.status === 200 || r.status === 405;
  } catch {
    return false;
  }
}

/**
 * Résout l'URL de base : cache 5 min, sinon détection (IP WiFi si dispo) + test des candidats.
 */
export async function resolveApiBaseUrl(): Promise<string> {
  // En Web, les probes cross-origin (ex: GET /auth/login/) sont bloqués par CORS.
  // On utilise donc une URL explicite (env) ou la prod.
  if (isWeb()) {
    const envUrl = (process.env.EXPO_PUBLIC_API_BASE_URL ?? '').trim();
    return envUrl || PRODUCTION_URL;
  }

  try {
    const cached = await AsyncStorage.getItem(CACHE_KEY);
    const tsStr = await AsyncStorage.getItem(CACHE_TS_KEY);
    const ts = tsStr ? parseInt(tsStr, 10) : 0;
    if (cached && Date.now() - ts < CACHE_TTL_MS) {
      return cached;
    }
  } catch {
    /* ignore */
  }

  const candidates = getCandidates();

  for (const url of candidates) {
    if (await tryBaseUrl(url)) {
      try {
        await AsyncStorage.setItem(CACHE_KEY, url);
        await AsyncStorage.setItem(CACHE_TS_KEY, String(Date.now()));
      } catch {
        /* ignore */
      }
      return url;
    }
  }

  try {
    await AsyncStorage.setItem(CACHE_KEY, PRODUCTION_URL);
    await AsyncStorage.setItem(CACHE_TS_KEY, String(Date.now()));
  } catch {
    /* ignore */
  }
  return PRODUCTION_URL;
}

export function invalidateApiBaseUrlCache(): Promise<void> {
  return Promise.all([
    AsyncStorage.removeItem(CACHE_KEY),
    AsyncStorage.removeItem(CACHE_TS_KEY),
  ]).then(() => undefined);
}
